Use Docker build secrets

omus opened the pull request:
Uses [`github-token-helper`](https://github.com/beacon-biosignals/github-token-helper) to avoid embedding the secrets in the image layers. Additional changes include:

- Conditionally set a private package registry via `--build-arg PRIVATE_REGISTRY_URL=...`
- Conditionally install Revise and Pprof via the build arg `ADD_UTILS`
- Skip installing PackageCompiler and attempting to make a system image if no packages are to be included in the system image
- Replace the default system image when using `create_sysimage`. This makes it easier to handle the previous point. The original system image can be restored via PackageCompilers normal mechanisms

TODO:
- [x] Document `PRIVATE_REGISTRY_URL` in README
- [x] Investigate if there is a secure way of passing in the GitHub token when starting the pod
- [x] Validate `julia_pod` changes (my instance I used to test this is having issues)

omus commented on add_me_to_your_PATH/Dockerfile.template:
```diff
@@ -43,18 +45,21 @@ RUN apt-get update && apt-get install -y \
 
 COPY sysimage.jl /JuliaProject/sysimage.jl
-RUN julia --project=/JuliaProject -e 'include("/JuliaProject/sysimage.jl")'
+RUN --mount=type=secret,id=github_token \
+    julia --project=/JuliaProject -e 'include("/JuliaProject/sysimage.jl")'
```
I've added the `--mount` to avoid having Julia complain it can't update the private registry for `RUN` instructions where this would occur. Note that only in this particular case it shouldn't be required as we're only adding PackageCompiler.jl which is in the public registry

ericphanson commented on add_me_to_your_PATH/sysimage.jl:
```diff
@@ -25,8 +28,8 @@ function main()
     println(packages)
-    create_sysimage(packages; sysimage_path="deps.so", cpu_target = "generic;sandybridge,-xsaveopt,clone_all;haswell,-rdrnd,base(1)")
-    Pkg.rm("PackageCompiler")
+    create_sysimage(packages; replace_default=true, cpu_target = "generic;sandybridge,-xsaveopt,clone_all;haswell,-rdrnd,base(1)")
```
Why replace the default? It seems confusing to me if julia_pod changes stuff outside of the pod environment (and can cause issues with versioning, eg you can’t trust your manifest if you’re using a custom sysimage built in a different environment)

kolia replied:
`sysimage.jl` should only ever get run in the context of the docker build, which is private to each manifest/julia env; this won't affect the julia environment outside of that context. In fact, julia_pod doesn't assume you have julia installed locally at all.

The main thing this does I'm guessing (@omus might have other motivations?) is make the `CMD julia` that gets called at the end of the `Dockerfile.template` simpler.

I think with this, we no longer need the `[[ -n "${JULIA_COMMAND}" ]]` block in the `julia_pod` script, since the julia command is just `julia` in all cases.

omus replied:
The motivation behind replacing the default system image was that it plays nicer when we fully skip making a system image (due to having no pinned packages). In that scenario the original `/deps.so` file would not exist causing any call of `julia --sysimage=/deps.so` to fail. I could copy the original system image to this location but replacement seems like a better option. Do note that PackageCompiler backs up the previous system image so you can always restore the original

ericphanson replied:
Ah ok, I didn’t realize that this is only inside the docker env. 👍

kolia commented on add_me_to_your_PATH/sysimage.jl:
```diff
@@ -13,6 +13,9 @@ if project.ispackage && !isfile(modulefile)
 end
 
+# Skip generating a system image when there are no dependencies
+isempty(project.dependencies) && exit(0)
```
This short-circuit should probably check whether the `packages` passed to `create_sysimage` in `main` is empty. There will almost always be deps, but not always pinned deps.

kolia replied:
Oh right, ok.

The only difference would be the case where the only dep is `PackageCompiler`, which is not worth changing what you have.

kolia reviewed (commented):
See comments.

It does seem like the `julia_pod` script needs to not call `julia` with the `--sysimage` flag anymore. Also, does `build_image` not need some `--secret` flags added in the `docker buildx build` commands?

I'd like to understand `github-token-helper` a bit more.

In general, how are you going about testing? We used to have a couple of example repos to test against in the private repo before this was made public. I've been testing against a couple of actual julia projects I use, plus passing in an `--image`, by hand.

omus commented:
> In general, how are you going about testing?

I was just testing with a manual `docker build` command. I'll need to make some updates to `julia_pod` before this can be merged

omus commented:
> I'd like to understand `github-token-helper` a bit more.

Some more details image in the README for https://github.com/beacon-biosignals/github-token-helper/. A quick overview is it uses the Git credential helper protocol which is supported by Git and Julia to inject credentials without user interaction. The way I've setup the token helper here is it injects the credentials specified in a secrets file or a environmental variable. In the case of Docker secrets they are only available when mounted which is ideal for a build. The environment variable is meant for runtime use to allow end users to use `docker run --env GITHUB_TOKEN` to pass in credentials

omus commented:
Fixed some issues with this PR. Should be functional at this point but I need to validate the changes. I've changed the PR to be a draft to reflect the current state.

omus commented:
Validated changes. Found a few bugs including a scenario where the pinned extractor fails due to their being no dependencies to review:

```
> [pinned-extractor 3/3] RUN julia --project=. -e 'using Pkg;                          pinned = [p.name for p in values(Pkg.dependencies()) if p.is_pinned];                          Pkg.rm([p for p in keys(Pkg.project().dependencies) if p ∉ pinned])':
#16 1.916 ERROR: rm requires at least one package
#16 2.744 Stacktrace:
#16 2.744  [1] pkgerror(msg::String)
#16 3.151    @ Pkg.Types /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Pkg/src/Types.jl:55
#16 3.226  [2] require_not_empty
#16 3.226    @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Pkg/src/API.jl:69 [inlined]
#16 3.226  [3] rm(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}; mode::Pkg.Types.PackageMode, kwargs::Base.Iterators.Pairs{Symbol, Base.PipeEndpoint, Tuple{Symbol}, NamedTuple{(:io,), Tuple{Base.PipeEndpoint}}})
#16 3.334    @ Pkg.API /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Pkg/src/API.jl:209
#16 3.334  [4] rm(pkgs::Vector{Pkg.Types.PackageSpec}; io::Base.PipeEndpoint, kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
#16 3.355    @ Pkg.API /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Pkg/src/API.jl:80
#16 3.355  [5] rm(pkgs::Vector{Pkg.Types.PackageSpec})
#16 3.357    @ Pkg.API /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Pkg/src/API.jl:78
#16 3.357  [6] #rm#31
#16 3.357    @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Pkg/src/API.jl:76 [inlined]
#16 3.357  [7] rm(pkgs::Vector{String})
#16 3.359    @ Pkg.API /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Pkg/src/API.jl:76
#16 3.359  [8] top-level scope
#16 3.359    @ none:1
------
error: failed to solve: rpc error: code = Unknown desc = executor failed running [/bin/sh -c julia --project=. -e 'using Pkg;                          pinned = [p.name for p in values(Pkg.dependencies()) if p.is_pinned];                          Pkg.rm([p for p in keys(Pkg.project().dependencies) if p ∉ pinned])']: exit code: 1
```

kolia commented on add_me_to_your_PATH/Dockerfile.template:
```diff
@@ -118,4 +126,4 @@ WORKDIR /JuliaProject
 RUN echo 'n=8; sleep(n); println("done waiting $n secs for logs/repl_history.jl to sync...")' >> /root/.julia/config/startup.jl
 
-CMD julia --sysimage=/deps.so
+CMD julia
```
The final `CMD julia` will not have access to the github token, so users won't be able to add private dependencies?

omus replied:
The original code would embed the `GITHUB_TOKEN` into the image which is why it was available for use during runtime. I was planning on punting this functionality to another PR but I've restored it here using K8s secrets to supply the pod with the token.

kolia reviewed (approved):
I think this is good as long as users can add private packages from within their `julia_pod` session, which doesn't seem to be the case anymore with this setup?
